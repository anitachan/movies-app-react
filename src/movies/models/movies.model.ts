export interface NowPlaying {
  page: number,
  results: Movie[],
  total_pages: number,
  total_results: number,
}

export interface Movie {
  popularity?: number;
  vote_count?: number;
  video?: boolean;
  poster_path: string;
  id: number;
  adult?: boolean;
  backdrop_path: string;
  original_language?: OriginalLanguage;
  original_title?: string;
  genre_ids?: number[];
  title: string;
  vote_average: number;
  overview: string;
  release_date?: string;
  poster_url?: string;
}

export enum OriginalLanguage {
  En = 'en',
  Es = 'es',
  Ko = 'ko',
}
