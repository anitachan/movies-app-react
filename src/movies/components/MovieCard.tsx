import { Box, Button, Rating, Stack, Typography } from '@mui/material';
import { Movie } from '../models/movies.model';
import { useNavigate } from 'react-router-dom';
import { MovieDetail } from '../models/movieDetail';

interface Props {
  movie: Movie | MovieDetail;
}

export const MovieCard = ({ movie }: Props) => {
  const url = process.env.REACT_APP_IMAGES_URL;
  const { backdrop_path, title, overview, vote_average } = movie;
  const navigate = useNavigate();

  const handleClick = () => {
    navigate(`/details/${movie.id}`);
  };

  return (
    <Box sx={{ position: 'relative' }}>
      <img
        src={`${url}${backdrop_path}?w=400&fit=crop&auto=format`}
        alt={title}
        loading="lazy"
        style={{
          width: '100%',
          height: 'auto',
          objectFit: 'cover',
          display: 'block',
          borderRadius: 4,
        }}
      />
      <Box
        sx={{
          bgcolor: 'rgba(0, 0, 0, 0.7)',
          color: 'white',
          p: 2,
          mt: '-4px',
          display: 'flex',
          flexDirection: 'row',
          justifyContent: 'space-between',
          alignItems: 'center',
          borderBottomLeftRadius: 4,
          borderBottomRightRadius: 4,
        }}
      >
        <Box>
          <Typography
            sx={{
              display: '-webkit-box',
              WebkitLineClamp: 1,
              WebkitBoxOrient: 'vertical',
              overflow: 'hidden',
              textOverflow: 'ellipsis',
            }}
            variant="subtitle1"
            fontWeight="bold"
          >
            {title}
          </Typography>
          <Typography
            sx={{
              display: '-webkit-box',
              WebkitLineClamp: 2,
              WebkitBoxOrient: 'vertical',
              overflow: 'hidden',
              textOverflow: 'ellipsis',
            }}
            variant="body2"
          >
            {overview}
          </Typography>
          <Stack direction="row" spacing={1} alignItems="center">
            <Rating value={vote_average} precision={0.1} max={10} readOnly />
            <Typography variant="body2">{vote_average.toPrecision(2)}</Typography>
          </Stack>
        </Box>
        <Button
          sx={{ marginLeft: '10px', whiteSpace: 'nowrap' }}
          variant="contained"
          size="small"
          color="primary"
          onClick={handleClick}
        >
          See More
        </Button>
      </Box>
    </Box>
  );
};
