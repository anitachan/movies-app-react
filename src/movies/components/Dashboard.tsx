import { Box } from '@mui/material';
import { useInfiniteScroll } from '../hooks/useInfiniteScroll';
import { Movie } from '../models/movies';
import { MovieCard } from './MovieCard';
import { MovieCardSkeleton } from './MovieCardSkeleton';

interface Props {
	movies: Movie[];
	loadMore: () => void;
}

export const Dashboard = ({ movies, loadMore }: Props) => {
	const loaderRef = useInfiniteScroll(loadMore);

	return (
		<>
			<Box
				sx={{
					display: 'grid',
					minHeight: '120vh',
					gridTemplateColumns: {
						xs: 'repeat(1, 1fr)',
						sm: 'repeat(2, 1fr)',
						md: 'repeat(3, 1fr)',
						lg: 'repeat(4, 1fr)',
					},
					gap: 2,
				}}
			>
				{movies.length > 0
					? movies.map((movie) => <MovieCard key={movie.id} movie={movie} />)
					: [...Array(9)].map((_, i) => <MovieCardSkeleton key={i} />)}
			</Box>
			{movies.length > 0 && <div ref={loaderRef} data-testid="loader" style={{ height: '60px' }} />}
		</>
	);
};
