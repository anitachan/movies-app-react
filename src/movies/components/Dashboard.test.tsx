import { render, screen } from '@testing-library/react';
import { Dashboard } from './Dashboard';
import { Movie, OriginalLanguage } from '../models/movies';
import { MovieCard as MockedMovieCard } from './MovieCard'; // ğŸ‘ˆ cambio aquÃ­

const mockMovies: Movie[] = [
	{
		adult: false,
		backdrop_path: '/t9nyF3r0WAlJ7Kr6xcRYI4jr9jm.jpg',
		genre_ids: [878, 28],
		id: 580489,
		original_language: OriginalLanguage.En,
		original_title: 'Venom: Let There Be Carnage',
		overview:
			'After finding a host body in investigative reporter Eddie Brock, the alien symbiote must face a new enemy, Carnage, the alter ego of serial killer Cletus Kasady.',
		popularity: 10352.991,
		poster_path: '/rjkmN1dniUHVYAtwuV3Tji7FsDO.jpg',
		release_date: '2021-09-30',
		title: 'Venom: Let There Be Carnage',
		video: false,
		vote_average: 7.2,
		vote_count: 518,
	},
	{
		adult: false,
		backdrop_path: '/8Y43POKjjKDGI9MH89NW0NAzzp8.jpg',
		genre_ids: [35, 28, 12, 878],
		id: 550988,
		original_language: OriginalLanguage.En,
		original_title: 'Free Guy',
		overview:
			'A bank teller called Guy realizes he is a background character in an open world video game called Free City that will soon go offline.',
		popularity: 5333.728,
		poster_path: '/xmbU4JTUm8rsdtn7Y3Fcm30GpeT.jpg',
		release_date: '2021-08-11',
		title: 'Free Guy',
		video: false,
		vote_average: 7.9,
		vote_count: 2611,
	},
	{
		adult: false,
		backdrop_path: '/aO9Nnv9GdwiPdkNO79TISlQ5bbG.jpg',
		genre_ids: [28, 12],
		id: 568620,
		original_language: OriginalLanguage.En,
		original_title: 'Snake Eyes: G.I. Joe Origins',
		overview:
			"After saving the life of their heir apparent, tenacious loner Snake Eyes is welcomed into an ancient Japanese clan called the Arashikage where he is taught the ways of the ninja warrior. But, when secrets from his past are revealed, Snake Eyes' honor and allegiance will be tested â€“ even if that means losing the trust of those closest to him.",
		popularity: 3101.51,
		poster_path: '/uIXF0sQGXOxQhbaEaKOi2VYlIL0.jpg',
		release_date: '2021-07-22',
		title: 'Snake Eyes: G.I. Joe Origins',
		video: false,
		vote_average: 6.9,
		vote_count: 635,
	},
];

const MovieCard = MockedMovieCard as jest.Mock;

jest.mock('../hooks/useInfiniteScroll', () => ({
	useInfiniteScroll: (cb: () => void) => {
		cb();
		return { current: null };
	},
}));
jest.mock('./MovieCard', () => ({
	__esModule: true,
	MovieCard: jest.fn(({ movie }) => <div data-testid={`movie-${movie.id}`} />),
}));

describe('Dashboard', () => {
	test('Dashboard matches snapshot', () => {
		const { container } = render(<Dashboard movies={mockMovies} loadMore={jest.fn()} />);
		expect(container).toMatchSnapshot();
	});

	test('should render dashboard', () => {
		render(<Dashboard movies={mockMovies} loadMore={jest.fn()} />);

		expect(MovieCard).toHaveBeenCalledTimes(3);

		expect(MovieCard.mock.calls[0][0]).toMatchObject({ movie: mockMovies[0] });

		expect(MovieCard.mock.calls[0][0]).toEqual(expect.objectContaining({ movie: mockMovies[0] }));
		expect(MovieCard.mock.calls[1][0]).toEqual(expect.objectContaining({ movie: mockMovies[1] }));
		expect(MovieCard.mock.calls[2][0]).toEqual(expect.objectContaining({ movie: mockMovies[2] }));
	});

	test('should render loader div if movies exist', () => {
		render(<Dashboard movies={mockMovies} loadMore={jest.fn()} />);

		expect(screen.getByTestId('loader')).toBeInTheDocument();
	});

	test('should not render loader div if no movies', () => {
		render(<Dashboard movies={[]} loadMore={jest.fn()} />);

		expect(screen.queryByTestId('loader')).not.toBeInTheDocument();
	});

	test('should call loadMore when loaderRef is triggered', () => {
		const loadMore = jest.fn();

		render(<Dashboard movies={mockMovies} loadMore={loadMore} />);

		expect(loadMore).toHaveBeenCalledTimes(1);
	});
});
