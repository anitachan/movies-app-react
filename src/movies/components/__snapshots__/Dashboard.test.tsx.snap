// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Dashboard matches snapshot 1`] = `
<div>
  <div
    class="MuiBox-root css-1cvwnaj"
  />
  <div
    data-testid="loader"
    style="height: 60px;"
  />
</div>
`;
